packages/react/src/runtime/ac.ts

```diff
@@ -0,0 +1,131 @@
+const UNDERSCORE_UNICODE = 95;
+
+/**
+ * This length includes the underscore,
+ * e.g. `"_1s4A"` would be a valid atomic group hash.
+ */
+const ATOMIC_GROUP_LENGTH = 5;
+
+/**
+ * Memoize the result of ac so if it is called with the same args, it returns immediately.
+ * Also, to prevent useless React rerenders
+ */
+const cache = new Map();
+
+/**
+ * `ac` returns an instance of AtomicGroups. The instance holds the knowledge of Atomic Group so we can chain `ac`.
+ * e.g. <div className={ax([ax(['_aaaa_b']), '_aaaa_c'])} />
+ */
+class AtomicGroups {
+  values: Record<string, string> | undefined;
+  cacheKey: string;
+  constructor(values: Record<string, string>, cacheKey: string) {
+    // An object stores the relation between Atomic group and actual class name
+    // e.g. { "aaaa": "a" } `aaaa` is the Atomic group and `a` is the actual class name
+    this.values = values;
```

Comment:
Just to confirm, do we assume that the keys and values of the Record `AtomicGroups.values` never themselves contain underscores? (e.g. "aaaa_bbbb" would never be a valid key in `AtomicGroups.values`)

I'm curious whether it is possible to get two different AtomicGroups with the same cache key

Reply:
"aaaa" is the atomic group. It's always 4 chars and doesn't not include underscores. You can see how it's generated at [here](https://github.com/atlassian-labs/compiled/blob/master/packages/css/src/plugins/atomicify-rules.ts#L30).